# Logistic regression

In [65]:
# Importing various libraries
import pandas as pd
import sklearn

In [66]:
df = pd.read_csv('../apple_quality.csv')
df
# The dataset cannot be shown in its entirety because it is too large to be displayed in a single page.
# In the polynomial regression model we only have ≈ 50 rows, here we have 4001 rows.

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0        0.0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1        1.0 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2        2.0 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3        3.0 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4        4.0  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   
...      ...       ...       ...        ...          ...        ...       ...   
3996  3996.0 -0.293118  1.949253  -0.204020    -0.640196   0.024523 -1.087900   
3997  3997.0 -2.634515 -2.138247  -2.440461     0.657223   2.199709  4.763859   
3998  3998.0 -4.008004 -1.779337   2.366397    -0.200329   2.161435  0.214488   
3999  3999.0  0.278540 -1.715505   0.121217    -1.154075   1.266677 -0.776571   
4000     NaN       NaN       NaN        NaN          NaN        NaN       NaN   

                                Acidity Quality  
0                          -0.491590483    good  
1                          -0.722809367    good  
2                           2.621636473     bad  
3                           0.790723217    good  
4                           0.501984036    good  
...                                 ...     ...  
3996                        1.854235285    good  
3997                       -1.334611391     bad  
3998                       -2.229719806    good  
3999                        1.599796456    good  
4000  Created_by_Nidula_Elgiriyewithana     NaN  

[4001 rows x 9 columns]

In [67]:
df.tail(1)
# We have an entire row with 'NaN's

A_id  Size  Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
4000   NaN   NaN     NaN        NaN          NaN        NaN       NaN   

                                Acidity Quality  
4000  Created_by_Nidula_Elgiriyewithana     NaN

In [68]:
# Removing this row using 'drop()' function
df.drop(4000, axis=0, inplace=True)

In [69]:
# Dividing the dataset into dependent and independant variables.
#   Dependant: Quality
#   Independant: All of the others except "A_id"

X = df.iloc[:, 1:-1].values
y = df.iloc[:, -1].values

In [70]:
# Since we have one column with a string we encode this column
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
le.fit(y)
y_enc = le.transform(y)

In [71]:
# Splitting the dataset into training and testing. Providing 80% for trainging and using a random state of 42
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y_enc, train_size=0.2, random_state=42)

In [72]:
# Instantiating the model of type logistic regression
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(random_state=42)

In [73]:
# Training the model on the test data i previously split
model.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [74]:
y_pred = model.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix, classification_report
cnf_matrix = confusion_matrix(y_test, y_pred)
report = classification_report(y_test, y_pred)

print("Confusion Matrix:\n" ,cnf_matrix)

print("\n\nClassification report:\n", report)

Confusion Matrix:
 [[1167  413]
 [ 404 1216]]


Classification report:
               precision    recall  f1-score   support

           0       0.74      0.74      0.74      1580
           1       0.75      0.75      0.75      1620

    accuracy                           0.74      3200
   macro avg       0.74      0.74      0.74      3200
weighted avg       0.74      0.74      0.74      3200



array([[1168,  413],
       [ 405, 1216]])

# Confusion metrics 
Tells us how many correct and wrong predictions we have. It is a 4x4 matrix
with these columns

|      | good | baad |
|------|------|------|
| good | x    | x    |
| baad | x    | x    |
 
 Here we have 1167 predsictions that predicted a good apple, and it was a good apple, and so forth...

# Classification report
**Precision**: Calculates the accuracy of the positive results
* (True positives)/(True Positives+False positives)

**Recall**: Identifies the models ability to correctly predict the result
* (True positives)/(True Positives+False negatives)

**F1-score**: Creates a mean of precision and recall. When we want to consider both
* 2 * (Precision*Recall)/(Precision+Recall)
     
They relate to the CNF matrix in the way that they use the different columns in the matrix to calculate the different values.